README.md
# alexa-mp3-validator

[![NPM Version](https://img.shields.io/npm/v/alexa-mp3-validator.svg)](https://www.npmjs.com/package/alexa-mp3-validator)
[![Build Status](https://travis-ci.org/tejashah88/alexa-mp3-validator.svg?branch=master)](https://travis-ci.org/tejashah88/alexa-mp3-validator)
[![Coverage Status](https://coveralls.io/repos/github/tejashah88/alexa-mp3-validator/badge.svg)](https://coveralls.io/github/tejashah88/alexa-mp3-validator)
[![dependencies Status](https://david-dm.org/tejashah88/alexa-mp3-validator/status.svg)](https://david-dm.org/tejashah88/alexa-mp3-validator)

A node module for validating your MP3 files to be used in your Amazon Alexa skill. If you are looking for a CLI version, check out [ssmlol](https://github.com/okofish/ssmlol). If you want to convert your MP3 files to be usable by Alexa, check out [this website](https://www.jovo.tech/audio-converter).

## Upgrading from < 1.0
If you are upgrading from < 1.0, please see [UPGRADING](UPGRADING.md) for migration instructions.

### Requirements of Alexa-ready MP3 Files

Every MP3 file you want Alexa to play must meet several basic requirements. They are as follows:
* A valid (MPEG version 2) MP3 file
* No longer than 90 seconds
* Encoded with a bit rate of exactly 48 kbps
* Encoded with a sample rate of exactly 16,000 Hz

## Usage

The module will check the requirements and return a Promise that returns true or false. An error will be thrown if the module is unable to read the file.

#### Function signature
```javascript
Boolean validateMP3(String filepath, optional Boolean throwOnValidationError = false)
```

#### Using with Promises
```javascript
const { validateMP3 } = require('alexa-mp3-validator');
validateMP3('path/to/file.mp3')
  .then(isValid => {
    // more logic goes here
  })
  .catch(err => {
    // error handling logic goes here
  })
```

#### Using with async/await
```javascript
const { validateMP3 } = require('alexa-mp3-validator');
(async () => {
  try {
    let isValid = await validateMP3('path/to/file.mp3');
    // more logic goes here
  } catch (err) {
    // error handling logic goes here
  }
})()
```

By default, if a validation error occurs, it will return false. By setting `throwOnValidationError` to true, all errors will throw.

#### Using with async/await and `throwOnValidationError = true`
```javascript
const { validateMP3, AudioValidationError } = require('alexa-mp3-validator');
(async () => {
  try {
    let isValid = await validateMP3('path/to/file.mp3', true);
    // more logic goes here
  } catch (err) {
    if (err instanceof AudioValidationError) {
      // validation error handling logic goes here
    } else {
      // other error handling logic goes here
    }
  }
})()
```

CHANGELOG.md
## Changelog

### 1.x (Next)
* Your commit here

### 1.0.3 (July/9/2019)
* Updated dependencies to resolve moderate DDoS security vulnerabilities. Travis CI integration now working for Node.js LTS versions 6 through 12. - [@tejashah88](https://github.com/tejashah88)

### 1.0.2 (June/7/2018)
* Dropped support for node v4 to meet compatibility with dependencies - [@tejashah88](https://github.com/tejashah88)

### 1.0.1 - Invalid

### 1.0.0 (June/7/2018)

* First official release - [@tejashah88](https://github.com/tejashah88)
  * If you were using this library prior to 1.0, please see the [UPGRADING](UPGRADING.md) document for migration instructions
* Added tests, travis CI, and code coverage from coveralls.io - [@tejashah88](https://github.com/tejashah88)
* Improved accuracy of detection of metadata by scanning all parts of the MP3 file and taking the most occurring values as the correct metadata representing the MP3 file - [@tejashah88](https://github.com/tejashah88)
* Changed `validateMP3` syntax and functionality to be mroe flexible with given inputs and error handling (see [UPGRADING](UPGRADING.md) for more details)


test/utils.test.js
'use strict';

const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const expect = require('chai').expect;

const utils = require('../src/utils');
const fileInfos = require('./fixtures/metadata.json');

describe('utils', function() {
  describe('#mode()', function() {
    it('should handle singular modes', function() {
      expect(utils.mode([1,2,2,3])).to.have.members([2]);
      expect(utils.mode([2,8,2,3,5,2,4,4,7,5,0,9,5,8,9,4,3,9,5,3,5,4,2,1,1])).to.have.members([5]);
    });

    it('should handle multiple modes', function() {
      expect(utils.mode([1,2,2,1])).to.have.members([1,2]);
      expect(utils.mode([2,8,2,3,5,2,4,4,7,5,0,1,5,1,9,4,3,9,5,3,5,4,2,1,1,1])).to.have.members([5,1]);
    });
  });

  describe('#getFileBuffer()', function() {
    for (const fileInfo of fileInfos) {
      describe(`with ${fileInfo.name}.mp3`, function() {
        const sampleFile = `./test/fixtures/${fileInfo.name}.mp3`;
        const promisedBuffer = utils.getFileBuffer(sampleFile);

        it('should return a fulfilled Promise', function() {
          return Promise.all([
            expect(promisedBuffer).to.be.a('promise'),
            expect(promisedBuffer).to.be.fulfilled
          ]);
        });

        it(`should eventually return a buffer whose length is ${fileInfo.size} bytes`, function() {
          return Promise.all([
            expect(promisedBuffer).to.eventually.be.an.instanceof(Buffer),
            expect(promisedBuffer).to.eventually.have.lengthOf(fileInfo.size)
          ]);
        });
      });
    }

    it('should fail when trying to read from an invalid path', function () {
      const sampleFile = `./test/fixtures/invalid.mp3`;
      const invalidPromisedBuffer = utils.getFileBuffer(sampleFile);

      it('should return a fulfilled Promise', function() {
        return Promise.all([
          expect(invalidPromisedBuffer).to.be.a('promise'),
          expect(invalidPromisedBuffer).to.be.rejected
        ]);
      });
    });
  });
});


test/mp3-metadata.test.js
'use strict';

const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const expect = require('chai').expect;

const getMp3Metadata = require('../src/mp3-metadata');
const utils = require('../src/utils');
const fileInfos = require('./fixtures/metadata.json');

describe('mp3-metadata', function() {
  describe('#getMp3Metadata()', function() {
    for (const fileInfo of fileInfos) {
      it(`${fileInfo.name}.mp3 should return the correct metadata`, function() {
        const sampleFile = `./test/fixtures/${fileInfo.name}.mp3`;
        const promisedMetadata = utils.getFileBuffer(sampleFile).then(buffer => getMp3Metadata(buffer));
        return expect(promisedMetadata).to.eventually.deep.equal(fileInfo.metadata);
      });
    }
  });
});


test/mp3-validator.test.js
'use strict';

const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const expect = require('chai').expect;

const { validateMP3 } = require('../src/mp3-validator');
const AudioValidationError = require('../src/AudioValidationError');
const fileInfos = require('./fixtures/metadata.json');

describe('mp3-validator', function() {
  describe('#validateMP3()', function() {
    describe('with throwOnValidationError = false', function() {
      for (const fileInfo of fileInfos) {
        it(`${fileInfo.name}.mp3 should ${fileInfo.alexaReady ? 'pass' : 'fail'} the MP3 validation`, function() {
          const sampleFile = `./test/fixtures/${fileInfo.name}.mp3`;
          return expect(validateMP3(sampleFile)).to.eventually.equal(fileInfo.alexaReady);
        });
      }
    });

    describe('with throwOnValidationError = true', function() {
      for (const fileInfo of fileInfos) {
        it(`${fileInfo.name}.mp3 should ${fileInfo.alexaReady ? 'pass' : 'fail'} the MP3 validation`, function() {
          const sampleFile = `./test/fixtures/${fileInfo.name}.mp3`;
          if (fileInfo.alexaReady)
            return expect(validateMP3(sampleFile, true)).to.eventually.equal(fileInfo.alexaReady);
          else
            return expect(validateMP3(sampleFile, true)).to.be.rejectedWith(AudioValidationError);
        });
      }
    });
  });
});


index.js
var fs = require('fs');
var utils = require('./utils');
var mp3Parser = require('mp3-parser');

module.exports = function validateMP3(filepath) {
  utils.isPlayableMP3(filepath);

  var mp3Info = mp3Parser.readFrameHeader(utils.dataViewFromFilePath(filepath));
  var mp3TimeLength = utils.calculateMP3Time(filepath, mp3Info.bitrate);

  utils.processID3Field(
    mp3Info.bitrate === 48,
    "Bitrate should be set to 48 kbps. Currently, it is %d kbps.",
    mp3Info.bitrate
  );

  utils.processID3Field(
    mp3Info.samplingRate === 16000,
    "Sampling Rate should be set to 16000 Hz or 16 kHz. Currently, it is %d Hz.",
    mp3Info.samplingRate
  );

  utils.processID3Field(
    mp3Info.mpegAudioVersionBits === '10',
    "MP3 Version should be 2.0 or higher."
  );

  utils.processID3Field(
    mp3TimeLength <= 90,
    "MP3 length should be no longer than 90 seconds. Currently, it is %d seconds.",
    utils.roundNPlaces(mp3TimeLength, 2)
  );
};

src/mp3-metadata.js
'use strict';

const Mp3Header = require('mp3-header').Mp3Header;
const getMP3Duration = require('get-mp3-duration');

const utils = require('./utils');

/*
 * This function is for getting the metadata of an MP3, specifically the version, bitrate and
 * sampling rate. It does this by selecting a header range to pull the potential metadata and
 * then skipping every 255 bits (or 0xFF) for a new header position. This is done because several
 * loosely conducted experiments revealed that relying on any one position for the header showed
 * contradictory results to the expected metadata. After collecting the array of possible versions,
 * bitrates, and sampling rates, it takes the most occurring value as the most probable value
 * for describing the property of that MP3 file.
 */
module.exports = function getMp3Metadata(buffer) {
  let pos = buffer.indexOf(0xFF);
  const versions = [], bitrates = [], samplingRates = [];

  while (pos >= 0) {
    const header = new Mp3Header(buffer.slice(pos, pos + 4));
    if (header.parsed && header.is_valid && header.mpeg_bitrate && header.mpeg_samplerate) {
      versions.push(header.mpeg_version);
      bitrates.push(header.mpeg_bitrate);
      samplingRates.push(header.mpeg_samplerate);
    }

    pos = buffer.indexOf(0xFF, pos + 1);
  }

  return {
    version: utils.mode(versions)[0],
    bitrate: utils.mode(bitrates)[0],
    samplingRate: utils.mode(samplingRates)[0],
    duration: getMP3Duration(buffer)
  };
};

src/mp3-validator.js
'use strict';

const roundTo = require('round-to');
const utils = require('./utils');
const AudioValidationError = require('./AudioValidationError');

const getMp3Metadata = require('./mp3-metadata');

function validateMP3(filepath, throwOnValidationError = false) {
  return utils.getFileBuffer(filepath)
    .then(buffer => {
      const metadata = getMp3Metadata(buffer);
      if (metadata.version !== 2)
        throw new AudioValidationError(`Expected MP3 Version to be 2.0 but got ${metadata.version.toFixed(1)}.`);
      if (metadata.samplingRate !== 16000)
        throw new AudioValidationError(`Expected sampling Rate to be 16,000 Hz but got ${metadata.samplingRate} Hz.`);
      if (metadata.bitrate !== 48000)
        throw new AudioValidationError(`Expected bitrate to be 48 kbps but got ${metadata.bitrate / 1000} kbps.`);
      if (metadata.duration > 90000)
        throw new AudioValidationError(`Expected duration to be no longer than 90 seconds but got ${roundTo(metadata.duration / 1000, 1)} seconds.`);

      return true;
    })
    .catch(err => {
      if (err instanceof AudioValidationError && !throwOnValidationError)
        return false;
      else
        throw err;
    });
}

module.exports = {
  validateMP3,
  AudioValidationError
};